### ISSUE TITLE
feat(training): export versioned ChessMamba artifacts for portable inference

### ISSUE THREAD

**Issue by johansolbakken** [BrageHK/bee-chess#61]
**Revised 2026-09-05** — broadening this from "export one ONNX file and
prove it runs" to the full contract this needs to actually be useful:
a versioned, hashable, self-verifying model bundle. See the revision
note at the bottom.

## Goal

The Linux/ROCm machine is allowed to be special for training. **The
artifact it produces must not be special.** Every promoted ChessMamba
checkpoint must export into a platform-independent bundle that runs on
macOS/Apple Silicon (and later, Rust Bee) without PyTorch or ROCm.

```text
Linux + AMD/ROCm
      │ train
      ▼
PyTorch checkpoint
      │ export + verify
      ▼
Portable model bundle
  ├── model.onnx
  ├── manifest.json
  └── parity data
      │
      ├──► M1 Mac / ONNX Runtime
      └──► Rust Bee / ONNX Runtime
```

## Why this is fixable (unchanged from the original diagnosis)

The model itself is not architecturally AMD-only:
- `ChessMamba`'s default scan backend (`mamba_core.py`'s
  `SCAN_BACKENDS["pscan"]`, from `mambapy.pscan`) is a Blelloch parallel
  scan built entirely from plain tensor ops — no custom CUDA/HIP kernel.
- The only genuinely ROCm/CUDA-only piece is the opt-in Triton fused
  kernel, not the default and not what `bridge/server.py` invokes.
- The rest of the forward path (`SpatialMixer`/`AttentionMixer`/policy
  and value heads) uses only gather/index_add/einsum/softmax over fixed
  64-square shapes — nothing device-specific.

So only the *training environment's dependency pinning* is ROCm-specific,
not the model. (That pinning bug — `[tool.uv.sources]` forcing the ROCm
wheel index for `torch`/`torchvision` unconditionally, unlike
`triton-rocm` which correctly has `sys_platform == 'linux'` — is real and
separately blocking local training-side dev on this machine, but is a
different, smaller fix; tracked separately, not folded into this issue.)

## Model bundle

```text
models/
└── chessmamba-0001/
    ├── model.onnx
    ├── manifest.json
    └── parity.json / parity.npz
```

`manifest.json` records at minimum (values below are illustrative —
generate from what the implementation actually produces):

```json
{
  "schema_version": 1,
  "model_type": "chessmamba",
  "model_version": "chessmamba-0001",
  "checkpoint_sha256": "...",
  "onnx_sha256": "...",
  "feature_set": "...",
  "perspective": "side_to_move",
  "policy_encoding": "from_to_64x64_no_underpromotion",
  "value_encoding": "categorical_cp_bins",
  "value_bins": 128,
  "cp_clip": 1000,
  "opset": 0,
  "dtype": "float32",
  "training_commit": "...",
  "training_run": "..."
}
```

`policy_encoding`/`value_encoding`/`value_bins`/`cp_clip`/`perspective`
should match exactly what #11/#66 already documented from the real
checkpoint (from/to-only policy, no underpromotion disambiguation;
hard-bin categorical value, not yet true HL-Gauss).

## Export command

One documented command, e.g.:

```bash
uv run python -m bee_training.chess_mamba.export \
    --checkpoint checkpoints/run-123/best.pt \
    --output models/chessmamba-0001/
```

It should: load the exact training architecture/config, load the
checkpoint, set eval mode, export both policy and value outputs, write
the manifest, compute artifact hashes, and fail loudly (not silently
produce a wrong/partial export) if something is unsupported. No
hand-editing Python source per checkpoint.

## Parity verification is mandatory

"An ONNX file was created" is not sufficient. Run a deterministic
position corpus (100-1000 positions) through both the PyTorch/ROCm model
and the ONNX Runtime model and compare:

- **Policy**: tensor shape, numerical tolerance, top-1 agreement, top-5
  ranking agreement.
- **Value**: tensor shape, bin logits/probabilities within tolerance,
  and the *decoded* centipawn evaluation within tolerance (softmax over
  bins → weighted sum of bin centers) — that decoded scalar is what
  Bee's `Evaluator` will actually consume.

Store enough of this golden data in the bundle that a consumer on
another machine can independently re-verify the artifact, rather than
trusting the exporter's own run. This immediately catches: bad export,
mismatched board encoding, wrong square indexing, wrong side-to-move
perspective, wrong value-bin decoding.

## macOS acceptance test

On Apple Silicon, without the training environment: install only
lightweight inference deps (`onnxruntime`, no torch/ROCm), load
`model.onnx`, encode the parity positions, run inference, and match the
bundle's recorded expected outputs within tolerance. No ROCm package, no
training checkpoint, no PyTorch dependency required for this test.

## Rust consumption contract (guarantee the shape, not the integration)

Rust integration itself stays in #66 — this issue only guarantees the
exported bundle is *sufficient* for it:

```text
Position → canonical feature encoder → ONNX inference → policy logits + value logits
```

Design the bundle so Bee can eventually support both a filesystem and an
embedded model source through the same loader/manifest validation:

```rust
enum ModelSource { File(PathBuf), Embedded(&'static [u8]) }
Model::load(source) // same validation, same semantics, either source
```

Prefer external files as the normal development path (`bee --model
./models/chessmamba-0001/model.onnx`); embedding via `include_bytes!` is
an optional release-packaging mode, not the default — otherwise every
new model forces a Rust rebuild and commits large binary diffs into the
normal source workflow.

## Artifact storage

Do not assume every promoted checkpoint gets committed straight to
normal Git history the way `training/checkpoints/main-dawg/latest.pt`
currently is (37MB, committed directly, no LFS). Decide per artifact
size: small enough may be fine in-repo; larger binaries want Git LFS or
an external artifact store; training checkpoints themselves (as opposed
to promoted portable bundles) should generally stay training artifacts,
not source-tree files. The manifest + hash is what makes an artifact
reproducible regardless of where its bytes physically live — not its
proximity to Rust source.

## Explicitly out of scope

This issue does **not**: retrain ChessMamba; change the policy/value
architecture; fix the underpromotion encoding; convert the value loss to
true HL-Gauss; integrate the model into alpha-beta; decide whether
neural value inference is fast enough for leaves; replace the existing
material evaluator. Those belong to #11/#66/#13 as already scoped.

## Definition of done

- [ ] One-command export from a real ChessMamba checkpoint to ONNX
- [ ] Both policy and value heads are exported
- [ ] Versioned `manifest.json` is generated
- [ ] Model and source-checkpoint hashes are recorded
- [ ] Deterministic parity corpus exists
- [ ] PyTorch ↔ ONNX policy parity is verified
- [ ] PyTorch ↔ ONNX value parity is verified
- [ ] Export fails clearly for unsupported/custom operators
- [ ] Exported model loads on Apple Silicon without ROCm
- [ ] M1 parity test passes without the training environment
- [ ] Batch-1 inference latency reported as an initial data point
- [ ] Docs cover the full train → export → verify → consume workflow
- [ ] Artifact format is sufficient for later Rust loading from either a
      filesystem path or embedded bytes

---
**Revision note:** the original version of this issue only asked to
prove `torch.onnx.export` works and produces a usable file. This
broadens it to the actual contract needed at the training/inference
boundary: a versioned, hashable, self-verifying bundle — because
"exported file exists" doesn't compose with #11's manifest requirements
or #66's need to independently verify an artifact without re-running the
exporter that produced it.

**Comment by johansolbakken** [comment]
:shipit:

**Comment by johansolbakken** [comment]
Deprioritizing: this shouldn't drive architecture ahead of #66 deciding what Bee-Mamba actually integrates as (value head for Bee's search, standalone searchless baseline, or both — see #66). Optimizing the standalone Python player's portability before that's settled risks solving the wrong problem.

ONNX may still end up valuable as the interchange/export format either way (#66 flags it as a plausible shared exporter target), so keeping this open, just behind #66 rather than ahead of it.

**Comment by johansolbakken** [comment]
Adding explicit success criteria — "an ONNX file exists" isn't enough,
and per #66's revised scope this now needs to cover both heads, not just
policy:

**Success is a parity corpus, not just export succeeding.** Build a
fixed set of positions (100-1000, e.g. sampled from existing self-play
records) and for each one compare:

- **Policy**: raw logit numerical difference, top-1 move agreement,
  top-5 ranking agreement (`policy_logits` reshaped to 64x64, see
  `FromToPolicyHead`).
- **Value**: bin-logit/probability agreement, and the *decoded* scalar
  (softmax over bins → weighted sum of bin-center cp values, using
  `cp_clip`/`n_value_bins` from the checkpoint's `TrainConfig`) — since
  that decoded scalar is what Bee's `Evaluator` will actually consume.

Both computed once from the PyTorch/ROCm checkpoint (on whatever machine
can run it) and once from the exported ONNX model (e.g. on this M1 Mac),
and compared within tolerance. That comparison — not the existence of a
`.onnx` file — is what proves the AMD dependency is actually broken.

This corpus is also the natural place to get an early read on inference
latency (median/p95 per-position, batch 1/8/32) ahead of #66/#13's more
formal benchmarking — cheap to capture while you're already running
every position through both models.

**Comment by johansolbakken** [comment]
Adding the missing piece: **how the bundle actually reaches the rest of
us**, not just what it contains. Concretely, once you've run the export
command and have `models/<name>/{model.onnx,manifest.json,parity.*}`:

**Where it goes:** don't commit promoted bundles as regular files
straight into `training/` the way `checkpoints/main-dawg/latest.pt` is
today (37MB, no LFS, and that's a *training checkpoint*, not even a
promoted portable bundle — see this issue's Artifact Storage section).
Pick one of:
- Git LFS, if we want it in-repo and versioned alongside code, or
- a shared location outside git entirely (e.g. a release/artifact
  upload, object storage, whatever's easiest for you to push to) with
  the manifest's hash as the thing that ties a repo reference back to
  the actual bytes.

Whichever you pick, the rest of us need is: **a manifest we can fetch or
read without needing your training environment**, plus a way to get the
`model.onnx` bytes it points to. If you're unsure which storage option is
easiest on your end, propose one in this issue and we'll agree on it —
that decision doesn't need to block you from finishing the exporter
itself.

**What "done" looks like from your side, concretely:**
1. Run the export command on a checkpoint you consider worth sharing.
2. Publish `model.onnx` + `manifest.json` + parity data wherever you and
   we agree (see above).
3. Tell us where — a path, a URL, a release tag, whatever it is — e.g.
   as a comment on this issue, or a short note in `lab/README.md`/
   `training/README.md` pointing at it.

That's the full loop: you don't need to wait for #66 (Rust integration)
or #11 (formal manifest schema finalization) to hand over a first
artifact — a working exporter + one published bundle is itself useful
for the rest of us to start poking at (parity-checking on M1, measuring
latency) even before Bee consumes it for real.

**Comment by johansolbakken** [comment]
@BrageHK tried to have Claude explain in this thread what we need to integrate the Mamba model into the engine. Can your Claude take a look?
